### PR TITLE
faster fragment finding implementation

### PR DESCRIPTION
**Pull request by richardjgowers**
Currently finding fragments is slow, this should be faster!

### CODE REVIEW

File: package/MDAnalysis/lib/_cutil.pyx [hunk 1]
```diff
@@ -310,4 +313,71 @@ cdef float _norm(float * a):
...
+        # Loop until fragment fully exploredt
+        while not frag_todo.empty():
+            for a in frag_todo:
+                frag_done.insert(a)
+                this_frag.push_back(a)
+                for b in bondmap[a]:
+                    if not frag_done.count(b):
+                        frag_next.insert(b)
+            # update todo list
+            frag_todo.clear()
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/_cutil.pyx]: this is kind of ugly, but I wanted to not change `frag_todo` whilst I was iterating over it, but maybe it's safe to do so?  Ideas welcome

File: package/MDAnalysis/lib/_cutil.pyx [hunk 2]
```diff
@@ -310,4 +313,71 @@ cdef float _norm(float * a):
...
+    frags = []
+
+    while not todo.empty():  # While not all nodes have been done
+        # Start a new fragment
+        frag_todo.clear()
+        frag_done.clear()
+        this_frag.clear()
+        # Grab a start point for next fragment
+        frag_todo.insert(deref(todo.begin()))
+        # Loop until fragment fully exploredt
```
Comment (jbarnoud) [inline package/MDAnalysis/lib/_cutil.pyx]: extra "t" at the end of the comment.

File: package/MDAnalysis/lib/_cutil.pyx (hunk 1 above)
Comment (jbarnoud) [reply package/MDAnalysis/lib/_cutil.pyx]: Since the swap is done in constant time, I am fine with this approach.

**Comment by jbarnoud** [comment]
The fact that the tests are broken may indicate that I missed something reading the code, though. The good thing is that there is no new windows specific failures,

File: package/MDAnalysis/lib/_cutil.pyx [hunk 3]
```diff
@@ -310,4 +313,71 @@ cdef float _norm(float * a):
...
     for n in range(3):
         result += a[n]*a[n]
-    return sqrt(result)
\ No newline at end of file
+    return sqrt(result)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def find_fragments(np.int64_t[:] atoms, int[:, :] bondlist):
```
[the remark this answers is not in the document]
Comment (richardjgowers) [reply package/MDAnalysis/lib/_cutil.pyx]: I actually put this in mdamath instead, but it has docs now

**Comment by richardjgowers** [comment]
@jbarnoud yeah the tests found that I was occasionally having duplicates in fragments, fixed now.  Tests, docs, benchmarks etc done, should be good to go

File: benchmarks/benchmarks/ag_methods.py [hunk 4]
```diff
@@ -319,3 +319,17 @@ def time_bond(self, num_atoms):
         Requires ag of size 2.
         """
         self.ag[:2].bond
+
+
+class FragmentFinding(object):
```
Comment (orbeckst) [inline benchmarks/benchmarks/ag_methods.py]: 👍 for added benchmark!

**Comment by jbarnoud** [comment]
AppVeyor is complaining about types. From what I understand, edges should be int64 instead of int32 in the tests.

**Comment by richardjgowers** [comment]
Ok that's weird, the appveyor github status should have been red?

**Comment by tylerjereddy** [comment]
appveyor looks red to me

**Comment by richardjgowers** [comment]
@MDAnalysis/coredevs this is ready

File: package/MDAnalysis/lib/_cutil.pyx [hunk 5]
```diff
@@ -312,3 +316,76 @@ cdef float _norm(float * a):
...
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def find_fragments(atoms, bondlist):
+    """Calculate distinct fragments from nodes and edges
+
+    Parameters
+    ----------
+    atoms : numpy.ndarray dtype int64
```
[the remark this answers is not in the document]
Comment (zemanj) [reply package/MDAnalysis/lib/_cutil.pyx]: Not sure if this should be tested?

**Comment by richardjgowers** [comment]
@zemanj thanks!  I was a little uncertain about modifying the set as I iterated over it, but this works nicely

**Comment by zemanj** [comment]
@richardjgowers I just made the docstring a little more verbose and made sure that `bondmap` is constructed from `bonds_view` instead of `bondlist`. In practice, this won't change anything but makes sure that *array_like* input also works correctly (now corresponding to the docstring). I also changed the inputs in the respective test to be nested lists instead of arrays for that reason. Using numpy arrays of correct shape and dtype should be covered by the tests in `core/test_fragments.py`.

**Comment by zemanj** [comment]
@richardjgowers BTW do you happen to know if it's possible in Cython to construct a `numpy.ndarray` directly from a `libcpp.set` without converting to a `libcpp.vector` first? I was unable to find anything useful on that topic...

**Comment by kain88-de** [comment]
There will be hardly a way do to this directly. The set uses a hash table
like data structure under the hood. How is the conversion to a vector done?

BTW using a normal for loop to fill a pre allocated numpy array should work
just fine for the conversion.

**Comment by zemanj** [comment]
@kain88-de Thanks, that's pretty much what I expected. I was just wondering whether something equivalent to the range constructor of the `std::vector` in C++ existed for numpy arrays in Cython...

**Comment by zemanj** [comment]
Ok, good to go. Merge?

**Comment by richardjgowers** [comment]
Yup, thanks for the help!

**Pull request merged**